Assets/Scripts/GameplayObjects/Obstacles/AsteroidFactory.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class AsteroidFactory : ObstacleBaseFactory
+{
+    #region Constants
+
+    private const string OBSTACLES_RESOURCE_NAME = "ScriptableObjects/Obstacle Params";
+
+    #endregion
+
+    #region Fields
+
+    private ObstacleParams _obstacleParams;
+
+    #endregion
+
+    #region Methods
+
+    public AsteroidFactory()
+    {
+        _obstacleParams = Resources.Load<ObstacleParams>(OBSTACLES_RESOURCE_NAME);
+    }
+
+    public override IObstacle Create()
+    {
+        var asteroid = (GameObject)Object.Instantiate(_obstacleParams.AsteroidPrefab);
+        return asteroid.GetComponent<AsteroidObstacle>();
```

Comment:
GetComponent doesn't necessarily get the object that you just created, see https://docs.unity3d.com/ScriptReference/GameObject.GetComponent.html (it might get you a different one that you created earlier)